### PR TITLE
Navbar: Implement Classes Option

### PR DESCRIPTION
**Pull request by cgack**
This includes the new navbar widget from the navbar review branch.

### CODE REVIEW

**Comment by cgack** [comment]
@arschmitz ready for review when you get a chance

File: js/widgets/navbar.backcompat.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+		// Browser globals
+		factory( jQuery );
+	}
+} )( function( $ ) {
+
+if ( $.mobileBackcompat !== false ) {
+	return $.widget( "mobile.navbar", $.mobile.navbar, {
+		_create: function() {
+			var self = this;
```
Comment (arschmitz) [inline js/widgets/navbar.backcompat.js]: Don't use self this break in some contexts self has special meaning `window.self === window` instead use `var that = this;`

File: js/widgets/navbar.js [hunk 2]
```diff
@@ -35,59 +35,112 @@ return $.widget( "mobile.navbar", {
...
-		grid: null
+		maxbutton: 5
 	},
 
 	_create: function() {
 
-		var $navbar = this.element,
-			$navbuttons = $navbar.find( "a, button" ),
-			iconpos = $navbuttons.filter( ":jqmData(icon)" ).length ? this.options.iconpos : undefined;
+		var self = this,
```
Comment (arschmitz) [inline js/widgets/navbar.js]: same as above don't use self

File: js/widgets/navbar.js [hunk 3]
```diff
@@ -35,59 +35,112 @@ return $.widget( "mobile.navbar", {
...
-				if ( theme ) {
-					classes += " ui-button-" + theme;
-				}
-				if ( icon ) {
-					classes += " ui-icon-" + icon + " ui-button-icon-" + iconpos;
-				}
-				$( this ).addClass( classes );
+			.jqmEnhanceable();
+
+		$.extend( self, {
```
Comment (arschmitz) [inline js/widgets/navbar.js]: no need to extend just do `this.{{thing}} = bar;` no need to merge anything here

File: js/widgets/navbar.js (hunk 2 above)
Comment (arschmitz) [inline js/widgets/navbar.js]: we removed onevar split this into individual statments

File: js/widgets/navbar.js [hunk 4]
```diff
@@ -35,59 +35,112 @@ return $.widget( "mobile.navbar", {
...
 
-		$navbar.delegate( "a", "vclick", function( /* event */ ) {
-			var activeBtn = $( this );
+	},
 
-			if ( !( activeBtn.hasClass( "ui-state-disabled" ) ||
+	_createNavRows: function() {
+		var rowCount, row, pos, buttonItem, overflowNav, navRow,
+			navItems = this.navbar.find( "li" ),
+			buttonCount = this.numButtons,
```
Comment (arschmitz) [inline js/widgets/navbar.js]: same here split var statements

File: js/widgets/navbar.js [hunk 5]
```diff
@@ -35,59 +35,112 @@ return $.widget( "mobile.navbar", {
...
+			}
+		}
+	},
 
-					// DEPRECATED as of 1.4.0 - remove after 1.4.0 release
-					// only ui-state-disabled should be present thereafter
-					activeBtn.hasClass( "ui-disabled" ) ||
-					activeBtn.hasClass( "ui-button-active" ) ) ) {
+	_makeNavButton: function( button, iconpos ) {
+		$( button ).button( { iconPosition: iconpos } );
```
Comment (arschmitz) [inline js/widgets/navbar.js]: i wonder should we use the widget here or css i lean towards css

File: js/widgets/navbar.js [hunk 6]
```diff
@@ -35,59 +35,112 @@ return $.widget( "mobile.navbar", {
...
-				} );
-			}
-		} );
+		self.navButtons = self.navbar.find( "a" );
+		self.numButtons = self.navButtons.length;
+
+		this._addClass( self.navbar, "ui-navbar" );
+		self.navbar.attr( "role", "navigation" )
+			.find( "ul" )
+			.jqmEnhanceable();
```
Comment (arschmitz) [inline js/widgets/navbar.js]: `jqmEnhanceable` is deprecated and should this be `find` or should it be `children` do we support multiple lists?

File: js/widgets/navbar.js (hunk 5 above)
Comment (cgack) [inline js/widgets/navbar.js]: we decided in the 10/1 meeting to stick to button widget for now

**Comment by cgack** [comment]
@arschmitz I made some changes. ready for another round when you get a chance

**Comment by dt1973** [comment]
Sorry to interject, but will this prevent jQM from inserting classes into my navbar elements? I'm using the structure only CSS, and right now I have to wrap all my stuff inside `<div></div>` to prevent this from happening.

**Comment by arschmitz** [comment]
@dt1973 can you open an issue explaining your issue with a jsbin example and we will be happy to take a look and consider this case.

File: js/widgets/navbar.js [hunk 7]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
 
 	options: {
 		iconpos: "top",
-		grid: null
+		maxbutton: 5
 	},
 
 	_create: function() {
+		var that = this;
```
Comment (arschmitz) [inline js/widgets/navbar.js]: remove not needed

File: js/widgets/navbar.js [hunk 8]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+			.find( "ul" );
 
-		var $navbar = this.element,
-			$navbuttons = $navbar.find( "a, button" ),
-			iconpos = $navbuttons.filter( ":jqmData(icon)" ).length ? this.options.iconpos : undefined;
+		that.navbar = navbar;
+		that.navButtons = navButtons;
+		that.numButtons = numButtons;
+		that.maxButton = maxButton;
+		that.iconpos = iconpos;
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this` on all of these

Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 9]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
-				if ( icon ) {
-					classes += " ui-icon-" + icon + " ui-button-icon-" + iconpos;
-				}
-				$( this ).addClass( classes );
+		 if ( numButtons <= maxButton ) {
+			navButtons.each( function() {
+				that._makeNavButton( this, iconpos );
 			} );
+		} else {
+			that._createNavRows();
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 10]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
-				} );
-			}
-		} );
+		that.navButtons = that.navbar.find( "a" );
+		that.numButtons = that.navButtons.length;
+
+		this._addClass( that.navbar, "ui-navbar" );
+		that.navbar.attr( "role", "navigation" )
+			.find( "ul" )
+			.jqmEnhanceable();
```
Comment (arschmitz) [inline js/widgets/navbar.js]: this is deprecated / removed.

File: js/widgets/navbar.morebutton.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            .appendTo( popupDiv );
+
+        // enhance buttons and move to new rows
+        for ( pos = 0; pos < buttonCount; pos++ ) {
+            buttonItem = navItems.eq( pos );
+            this._makeNavButton( buttonItem.find( "a" ), iconpos );
+            if ( pos + 1 === maxButton ) {
+
+                moreButton = $( "<li></li>" ).append( $( "<a></a>" )
+                                    .attr( "href", "#" + id )
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: this should not go through nav you should do `.popup( "open" )`

Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: also make this a `button`

File: js/widgets/navbar.morebutton.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+                popupNav.append( buttonItem );
+            }
+            popupNav.listview();
+
+        }
+        popupDiv.appendTo( this.navbar );
+        popupDiv.popup();
+
+        $.extend( this, {
+            moreButton: moreButton,
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: just do `this.moreButton = moreButton;` etc

File: js/widgets/navbar.morebutton.js [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+            popupDiv: popupDiv
+        } );
+    },
+
+    refresh: function() {
+        var newitems,
+            self = this,
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: dont use self

File: js/widgets/navbar.morebutton.js [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+            iconpos = self.iconpos;
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: onevar

File: js/widgets/navbar.morebutton.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        if ( !self.options.morebutton ) {
+          self._super();
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+          return;
+        }
+
+        if ( self.popupDiv ) {
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.js [hunk 17]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+
+		 if ( that.numButtons <= that.maxButton ) {
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 18]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+			that.navButtons.each( function() {
+				that._makeNavButton( this, that.iconpos );
+			} );
+		} else {
+			that._createNavRows();
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 19]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+	refresh: function() {
+		var that = this;
 
-				// The code below is a workaround to fix #1181
-				$( document ).one( "pagehide", function() {
-					activeBtn.removeClass( "ui-button-active" );
-				} );
-			}
-		} );
+		that.navButtons = that.navbar.find( "a" );
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 20]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+		that.numButtons = that.navButtons.length;
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 21]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+
+		this._addClass( that.navbar, "ui-navbar" );
+		that.navbar.attr( "role", "navigation" )
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 22]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+			.find( "ul" )
+			.jqmEnhanceable();
+
+		 if ( that.numButtons <= that.maxButton ) {
+			that.navButtons.each( function() {
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 23]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+		} else {
+			that._createNavRows();
+		}
+	},
+
+	_destroy: function() {
+		var navrows;
+		var that = this;
+
+		if ( that.numButtons > that.maxButton ) {
```
Comment (arschmitz) [inline js/widgets/navbar.js]: use `this`

File: js/widgets/navbar.js [hunk 24]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+			navrows = that.navbar.find( ".ui-navbar-row li" ).detach();
```
Comment (arschmitz) [inline js/widgets/navbar.js]: `this`

File: js/widgets/navbar.js [hunk 25]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+			$( ".ui-navbar-row" ).remove();
+			that.navbar.find( "ul" ).append( navrows );
```
Comment (arschmitz) [inline js/widgets/navbar.js]: `this`

File: js/widgets/navbar.js [hunk 26]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
+		}
+
+		this._removeClass( that.navbar, "ui-navbar" );
```
Comment (arschmitz) [inline js/widgets/navbar.js]: `this.navbar`

File: js/widgets/navbar.js [hunk 27]
```diff
@@ -35,59 +35,113 @@ return $.widget( "mobile.navbar", {
...
 
-		// Buttons in the navbar with ui-state-persist class should regain their active state before page show
-		$navbar.closest( ".ui-page" ).bind( "pagebeforeshow", function() {
-			$navbuttons.filter( ".ui-state-persist" ).addClass( "ui-button-active" );
+		that.navButtons.each( function() {
```
Comment (arschmitz) [inline js/widgets/navbar.js]: `this`

File: js/widgets/navbar.morebutton.js [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+        factory( jQuery );
+    }
+} )( function( $ ) {
+
+return $.widget( "mobile.navbar", $.mobile.navbar, {
+
+    options: {
+        morebutton: false,
+        morebuttontext: "...",
+        morebuttoniconpos: "top",
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: do we really want the default of `top` here

File: js/widgets/navbar.morebutton.js [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+        morebuttonicon: null
+    },
+
+    _create: function() {
+
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: remove blank line

File: js/widgets/navbar.morebutton.js [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+        this._super();
+        if ( this.options.morebutton  && this.numButtons > this.maxButton ) {
+            this._createNavPopup();
+        }
+    },
+
+    _id: function() {
+        return ( this.element.attr( "id" ) || ( this.widgetName + this.uuid ) );
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: remove unneeded parens

File: js/widgets/navbar.morebutton.js [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+    _createNavRows: function() {
+        if ( this.options.morebutton ) {
+            return;
+        }
+
+        this._super();
+    },
+
+    _createNavPopup: function() {
+        var popupDiv, popupNav, moreButton, pos, buttonItem, id,
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: onevar

File: js/widgets/navbar.morebutton.js [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+            navItems = this.navbar.find( "li" ),
+            buttonCount = navItems.length,
+            maxButton = this.maxButton,
+            iconpos = this.iconpos,
+            icon = this.options.morebuttonicon;
+
+        id = this._id() + "-popup";
+
+        popupDiv = $( "<div class='ui-navbar-popup' id='" + id + "'></div>" );
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: classes option

File: js/widgets/navbar.morebutton.js [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+        popupNav = $( "<ul class='ui-navbar-popupnav'>" )
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: classes option

File: js/widgets/navbar.morebutton.js (hunk 31 above)
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: with onevar gone declare pos in the for loop directly

File: js/widgets/navbar.morebutton.js [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+              self._makeNavButton( this, iconpos );
+            } );
+            newitems.appendTo( self.popupDiv.find( "ul" ) );
+        }
+        self._createNavPopup();
+    },
+
+    _destroy: function() {
+        var navitems,
+            self = this;
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: dont use self

File: js/widgets/navbar.morebutton.js [hunk 35]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        if ( self.popupDiv ) {
+            newitems = self.moreButton.parent().nextAll();
+            newitems.find( "a" ).each( function() {
+              self._makeNavButton( this, iconpos );
+            } );
+            newitems.appendTo( self.popupDiv.find( "ul" ) );
+        }
+        self._createNavPopup();
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 36]
```diff
@@ -0,0 +1,151 @@
...
+          self._super();
+          return;
+        }
+
+        if ( self.popupDiv ) {
+            newitems = self.moreButton.parent().nextAll();
+            newitems.find( "a" ).each( function() {
+              self._makeNavButton( this, iconpos );
+            } );
+            newitems.appendTo( self.popupDiv.find( "ul" ) );
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 37]
```diff
@@ -0,0 +1,151 @@
...
+        var newitems,
+            self = this,
+            iconpos = self.iconpos;
+        if ( !self.options.morebutton ) {
+          self._super();
+          return;
+        }
+
+        if ( self.popupDiv ) {
+            newitems = self.moreButton.parent().nextAll();
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 38]
```diff
@@ -0,0 +1,151 @@
...
+            newitems.appendTo( self.popupDiv.find( "ul" ) );
+        }
+        self._createNavPopup();
+    },
+
+    _destroy: function() {
+        var navitems,
+            self = this;
+
+        if ( !self.options.morebutton ) {
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 39]
```diff
@@ -0,0 +1,151 @@
...
+            self._super();
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 40]
```diff
@@ -0,0 +1,151 @@
...
+            return;
+        }
+
+        if ( self.popupDiv ) {
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: use `this`

File: js/widgets/navbar.morebutton.js [hunk 41]
```diff
@@ -0,0 +1,151 @@
...
+            navitems = self.popupDiv.find( "li" ).detach();
+            self.popupDiv.remove();
+            self.moreButton.parent().remove();
+            self.navbar.find( "ul" ).append( navitems );
+            self.navbar.removeClass( "ui-navbar" );
+            self.navButtons = self.navbar.find( "a" );
+            self.navButtons.each( function() {
```
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: all of the above should be `this`

File: js/widgets/navbar.morebutton.js (hunk 34 above)
Comment (arschmitz) [inline js/widgets/navbar.morebutton.js]: no context switches here remove completely

File: tests/unit/navbar/backcompat_tests.html [hunk 42]
```diff
@@ -0,0 +1,108 @@
...
+	<script src="../../../external/requirejs/require.js"></script>
+	<script src="../../../js/requirejs.config.js"></script>
+	<script src="../../../js/jquery.tag.inserter.js"></script>
+	<script src="../../jquery.setNameSpace.js"></script>
+	<script src="../../../tests/jquery.testHelper.js"></script>
+	<script src="../../../external/qunit/qunit.js"></script>
+	<script src="../../../external/qunit-assert-classes/qunit-assert-classes.js"></script>
+	<script>
+		$.testHelper.asyncLoad([
+			[ "widgets/page" ],
```
Comment (arschmitz) [inline tests/unit/navbar/backcompat_tests.html]: remove this is part of init

File: tests/unit/navbar/index.html [hunk 43]
```diff
@@ -11,11 +11,13 @@
 	<script src="../../jquery.setNameSpace.js"></script>
 	<script src="../../../tests/jquery.testHelper.js"></script>
 	<script src="../../../external/qunit/qunit.js"></script>
+	<script src="../../../external/qunit-assert-classes/qunit-assert-classes.js"></script>
```
Comment (arschmitz) [inline tests/unit/navbar/index.html]: disable backcompat and replace `nstest` with `ui`

File: tests/unit/navbar/navbar_backcompat.js [hunk 44]
```diff
@@ -0,0 +1,20 @@
+/*
+ * mobile navbar unit tests
+ */
+( function( QUnit, $ ) {
+QUnit.test( "navbar button gets active button class when clicked", function( assert ) {
+	var link = $( "#disabled-button-click a:not(.ui-state-disabled)" ).first();
+
+	link.click();
+	assert.ok( link.hasClass( "ui-button-active" ), "link has active button class" );
```
Comment (arschmitz) [inline tests/unit/navbar/navbar_backcompat.js]: uses `assert.hasClasses`

File: tests/unit/navbar/navbar_backcompat.js [hunk 45]
```diff
@@ -0,0 +1,20 @@
...
+} );
+
+QUnit.test( "disabled navbar button doesn't add active button class when clicked",
+    function( assert ) {
+        var link = $( "#disabled-button-click a.ui-disabled" ).first();
+
+        link.click();
+        assert.ok( !link.hasClass( "ui-button-active" ), "link doesn't have active button class" );
```
Comment (arschmitz) [inline tests/unit/navbar/navbar_backcompat.js]: use `assert.lacksClass`

File: tests/unit/navbar/navbar_core.js [hunk 46]
```diff
@@ -2,41 +2,87 @@
  * mobile navbar unit tests
  */
 ( function( QUnit, $ ) {
-QUnit.test( "navbar button gets active button class when clicked", function( assert ) {
-	var link = $( "#disabled-button-click a:not(.ui-disabled)" ).first();
 
-	link.click();
-	assert.ok( link.hasClass( "ui-button-active" ), "link has active button class" );
+QUnit.test( "classes are correctly assigned", function( assert ) {
+	var $ul = $( "#enhanced-classes" ),
```
Comment (arschmitz) [inline tests/unit/navbar/navbar_core.js]: hungarian notation

File: tests/unit/navbar/navbar_core.js [hunk 47]
```diff
@@ -2,41 +2,87 @@
...
+		r = $ul.find( "li" ).eq( 0 ).find( "a" ).find( "span" ),
```
Comment (arschmitz) [inline tests/unit/navbar/navbar_core.js]: `.find( "a span" )`

**Comment by arschmitz** [comment]
@cgack done with review nothing to major mostly style related

File: js/widgets/navbar.morebutton.js (hunk 28 above)
Comment (cgack) [inline js/widgets/navbar.morebutton.js]: Yeah, the default in navbar is top, so this matches that.

**Comment by cgack** [comment]
@arschmitz Ready for another round of review

**Comment by cgack** [comment]
@arschmitz I updated the more button to actually be a button. Now ready for review

**Comment by arschmitz** [comment]
:+1:

**Comment by cgack** [comment]
Closing with f3bd7c4811ec3177417e86d47ffbf68de688a1ef landing on 1.5-dev

**Pull request closed without merging**